### PR TITLE
IGNITE-14864: Schema update. Merge multiple converters stages.

### PR DESCRIPTION
**Pull request by AMashenkov**

### CODE REVIEW

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/SchemaChangeTableViewTest.java [hunk 1]
```diff
@@ -113,4 +117,220 @@ public void testAddNewColumn() {
...
+            assertEquals(222, (Integer)tbl.get(keyTuple2).value("valInt"));
+            assertEquals("newDefault", tbl.get(keyTuple2).value("valStr"));
+            assertEquals("newDefault", tbl.get(keyTuple2).value("val"));
+
+            Tuple keyTuple3 = tbl.tupleBuilder().set("key", 3L).build();
+
+            assertEquals(333, (Integer)tbl.get(keyTuple3).value("valInt"));
+            assertEquals("brandNewDefault", tbl.get(keyTuple3).value("valStr"));
+            assertEquals("brandNewDefault", tbl.get(keyTuple3).value("val"));
+
```
Comment (tledkov-gridgain) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/SchemaChangeTableViewTest.java:333]: Please remove empty line

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/SchemaChangeKVViewTest.java [hunk 2]
```diff
@@ -116,4 +120,188 @@ public void testAddNewColumn() {
...
+            assertEquals(222, (Integer)vkView.get(keyTuple2).value("valInt"));
+            assertEquals("newDefault", vkView.get(keyTuple2).value("valStr"));
+            assertEquals("newDefault", vkView.get(keyTuple2).value("val"));
+
+            Tuple keyTuple3 = vkView.tupleBuilder().set("key", 3L).build();
+
+            assertEquals(333, (Integer)vkView.get(keyTuple3).value("valInt"));
+            assertEquals("brandNewDefault", vkView.get(keyTuple3).value("valStr"));
+            assertEquals("brandNewDefault", vkView.get(keyTuple3).value("val"));
+
```
Comment (tledkov-gridgain) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/SchemaChangeKVViewTest.java:304]: Remove empty line

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/registry/SchemaRegistryImpl.java [hunk 3]
```diff
@@ -106,9 +114,37 @@ public SchemaRegistryImpl(int initialVer, Function<Integer, SchemaDescriptor> hi
...
+
+        if ((mapping = mappingCache.get(mappingKey)) != null)
+            return mapping;
+
+        mapping = schema(rowSchema.version() + 1).columnMapping();
+
+        for (int i = rowSchema.version() + 2; i <= curSchema.version(); i++)
+            mapping = ColumnMapping.mergeMapping(mapping, schema(i));
+
+        mappingCache.putIfAbsent(mappingKey, mapping);
```
Comment (korlov42) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/registry/SchemaRegistryImpl.java:145]: we need to remove corresponding mappings when old schema is dropped

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/registry/SchemaRegistryImpl.java [hunk 4]
```diff
@@ -106,9 +114,37 @@ public SchemaRegistryImpl(int initialVer, Function<Integer, SchemaDescriptor> hi
...
+
+        return new UpgradingRowAdapter(curSchema, row, mapping);
+    }
+
+    /**
+     * @param rowSchema Row schema.
+     * @param curSchema Target schema.
+     * @return Column mapper for target schema.
+     */
+    public ColumnMapper resolveMapping(SchemaDescriptor rowSchema, SchemaDescriptor curSchema) {
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/registry/SchemaRegistryImpl.java:129]: to make this method complete it worth to handle case when both versions are equal:
```
if (curSchema.version() == rowSchema.version())
    return ColumnMapping.identityMapping();
```

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/AbstractSchemaChangeTest.java [hunk 5]
```diff
@@ -93,6 +98,96 @@ void afterEach() throws Exception {
...
+     * Check unsupported column type change.
+     */
+    @Disabled("https://issues.apache.org/jira/browse/IGNITE-15056")
+    @Test
+    public void testChangeColumnType() {
+        List<Ignite> grid = startGrid();
+
+        createTable(grid);
+
+        SchemaBuilders.column("valInt", ColumnType.string()).build();
```
Comment (korlov42) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/AbstractSchemaChangeTest.java:111]: here and below result of this line is ignored

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/registry/SchemaRegistryImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (AMashenkov) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/registry/SchemaRegistryImpl.java]: I think, in this case, we should never get here.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/registry/SchemaRegistryImpl.java [hunk 6]
```diff
@@ -145,6 +181,7 @@ public void onSchemaDropped(int ver) {
         if (ver >= lastVer || ver <= 0 || schemaCache.keySet().first() < ver)
             throw new SchemaRegistryException("Incorrect schema version to clean up to: " + ver);
 
-        schemaCache.remove(ver);
+        if (schemaCache.remove(ver) != null)
+            mappingCache.keySet().stream().filter(k -> (k & 0xFFFFFFFF) == ver).forEach(mappingCache::remove);
```
Comment (korlov42) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/registry/SchemaRegistryImpl.java:185]:
```suggestion
           mappingCache.keySet().removeIf(k -> (k & 0xFFFFFFFFL) == ver);
```

Please note the `L` at the end of the integer literal. Without it cache won't be cleaned.

**Review by korlov42 — APPROVED** [review]

**Pull request merged**